ipa-core/src/test_fixture/event_gen.rs

```diff
@@ -166,15 +170,21 @@ impl<R: Rng> EventGenerator<R> {
             config,
             rng,
             users: vec![],
-            used: HashSet::new(),
+            used_ids: HashSet::new(),
         }
     }
 
-    fn gen_event(&mut self, user_id: UserId) -> TestRawDataRecord {
-        // Generate a new random timestamp between [0..`max_timestamp`).
-        // This means the generated events must be sorted by timestamp before being
-        // fed into the IPA protocols.
-        let current_ts = self.rng.gen_range(0..self.config.max_timestamp.get());
+    fn gen_event(&mut self, idx: usize) -> TestRawDataRecord {
+        let user_id = self.users[idx].user_id;
+
+        // Generate a new random timestamp between [0..`max_timestamp`) and distinct from
+        // already-used timestamps.
+        let current_ts = loop {
```

Comment:
this has some potential to stall, so maybe we could check if `used_timestamp` size == max_timestamp?